Update event source guidelines (#2611)
openhab/openhab-docs: developers/utils/events.md

If an event is passed through multiple components before being processed, it should preserve the original source, and append (separated by `=>`) another component for itself.
A few concrete examples:

- `org.openhab.ui=>org.openhab.core.io.rest$admin` would mean the event was initiated by a user named `admin` via Main UI (through the REST API).
- `org.openhab.ui.basic$default:03=>org.openhab.core.io.rest` would mean the event was initiated via the `default` sitemap in Basic UI, on the third page (through the REST API, and unauthenticated).
- `org.openhab.io.homekit$1467397f-c2e7-4b15-a7dc-315331ced2db` would mean the event was initiated from the HomeKit addon by a user identified by a UUID.
- `org.openhab.ui.basic$default:03=>org.openhab.android$my-phone=>org.openhab.io.openhabcloud$user@gmail.com=>org.openhab.core.io.rest` would mean the event was initiated by using a sitemap in the Android openHAB app, and proxied through myopenhab.org using the user `user@gmail.com` (and finally through the REST API, but without a user).
- `org.openhab.core.io.console` would mean the event was sent via the Karaf console.

> - `org.openhab.automation.jrubyscripting$rule:virtual_lighting:98` would mean the event was sent by a JRuby rule with ID `virtual_lighting:98`.

Author: A rule written with the JRuby helper library is able to populate the Rule ID. DSL rules have no way to do so, so are handled generically with the static source of `org.openhab.core.automation`. The source comes from https://github.com/openhab/openhab-jruby/blob/23a04794d81a90686badb3367a55a1b3c70fcdec/lib/openhab/core/items/item.rb#L644, thus the bundle `org.openhab.automation.jrubyscripting`, and not `org.openhab.core.automation`.

BUT... I'll also add an example for a DSL rule, and clarify wording around rule/action.

> - `org.openhab.automation.jrubyscripting$rule:virtual_lighting:98` would mean the event was sent by a JRuby rule with ID `virtual_lighting:98`.

Reviewer: > Okay, let me rephrase - `DSL rules have no way to do so. Any other language may be able to do so, but to my knowledge has not done so as of writing.`
>
> I'm not sure what you're talking about for SimpleRule? The source is sent by the BusEventActionService, which can be called by any sort of module (be it action, condition, whatever).

I think you're missing my point - I'm just saying that framing it as if it's DSL that "uses `org.openhab.automation`" would be wrong. You know better than me how the JRuby helper lib works, I'm not even sure exactly under which circumstances these rule-created events occur, but are you saying that any event generated using the JRuby helper library will apply this "custom" source? Even if it's just one of several actions in a rule? I assume that it's also possible to send events from JRuby that doesn't use the function in the helper library, so what this source would actually identify is the use of the helper library, not the JRuby script engine.

> Knowing a more precise source of a command will be extremely useful for end users debugging their rules when looking at logs. It can also help for rule implementation preventing infinite loops - I have a few rules that check the source of a command, and do nothing if it's from a particular other rule.

I agree that knowing what rule that triggers it is useful, but I'd think that the rule UID is much more relevant for that than e.g. `org.openhab.automation.jrubyscripting`.

What about rules that use different languages or a language in combination with built-in conditions and actions. Should they also be considered e.g. "a JRuby rule"? I assume that it's the helper lib in use by the action that generates the event that will actually determine what form the source will have, I'm just asking if this is useful. I would think that, what would be relevant to know regarding source for a rule generated event is the fact that the source *is* a rule, and the rule UID.

> - `org.openhab.automation.jrubyscripting$rule:virtual_lighting:98` would mean the event was sent by a JRuby rule with ID `virtual_lighting:98`.

Reviewer 2: > It's not that simple, the special case is those rules created by the helper library, not the DSL rules. A rule created in the UI, or by a YAML file, or any other generic way, will behave like "a DSL rule" - that also applies if the rule uses JRuby.

JS Scripting actually sets the source based on ruleUID or filename, regardless whether you use file-based rules or ScriptAction/ScriptConditions. This is done by the helper library when calling `items.getItem(...).sendCommand(...)` or `items.getItem(...).postUpdate(...)`.
If I wanted to, I could implement that in the add-on itself, but that would be way more complicated.

> - `org.openhab.automation.jrubyscripting$rule:virtual_lighting:98` would mean the event was sent by a JRuby rule with ID `virtual_lighting:98`.

Reviewer: Yes, my idea was that it would be more uniform if all events originating from rules had a form similar to `org.openhab.core.rule$<ruleUid>`, but you're correct that it wouldn't work anyway, because the rule UID probably can't be inferred if not explicitly provided by the caller. Which makes "my idea" pointless - so please go ahead as planned 😉

The event source identifies the sender.
Not all senders will set the source.
The source should be structured, consisting of a series of concatenated components (separated by `=>` to indicate delegation), and each component consists of a bundle name with an optional actor name (separated by `$`).

> Bundle names cannot contain either the delegation separator, or the actor separator.

Reviewer 2: ```suggestion
Bundle names can contain neither the delegation separator, nor the actor separator.
```

Personally I think that's easier to read than double negation.
